les/vflux/server/clientpool.go

```diff
@@ -0,0 +1,375 @@
+}
+
+// clientPeer represents a peer in the client pool. None of the callbacks should block.
+type clientPeer interface {
+	Node() *enode.Node
+	FreeClientId() string                         // unique id for non-priority clients (typically a prefix of the network address)
+	InactiveTimeout() time.Duration               // disconnection timeout for inactive non-priority peers
+	UpdateCapacity(newCap uint64, requested bool) // signals a capacity update (requested is true if it is a result of a SetCapacity call on the given peer
+	Disconnect()                                  // initiates disconnection (Unregister should always be called)
+}
+
+type clientPeerInstance struct{ clientPeer } // the NodeStateMachine type system needs this wrapper
+
+// NewClientPool creates a new client pool
+func NewClientPool(balanceDb ethdb.KeyValueStore, minCap uint64, connectedBias time.Duration, clock mclock.Clock, synced func() bool) *ClientPool {
+	ns := nodestate.NewNodeStateMachine(nil, nil, clock, serverSetup)
+	cp := &ClientPool{
+		ns:             ns,
+		balanceTracker: newBalanceTracker(ns, btSetup, balanceDb, clock, &utils.Expirer{}, &utils.Expirer{}),
+		priorityPool:   newPriorityPool(ns, ppSetup, clock, minCap, connectedBias, 4),
+		clock:          clock,
+		minCap:         minCap,
+		connectedBias:  connectedBias,
+		synced:         synced,
+	}
+
+	ns.SubscribeState(nodestate.MergeFlags(ppSetup.activeFlag, ppSetup.inactiveFlag, btSetup.priorityFlag), func(node *enode.Node, oldState, newState nodestate.Flags) {
+		if newState.Equals(ppSetup.inactiveFlag) {
+			// set timeout for non-priority inactive client
+			var timeout time.Duration
+			if c, ok := ns.GetField(node, clientField).(clientPeer); ok {
+				timeout = c.InactiveTimeout()
+			}
+			if timeout > 0 {
+				ns.AddTimeout(node, ppSetup.inactiveFlag, timeout)
+			} else {
+				// Note: if capacity is immediately available then priorityPool will set the active
+				// flag simultaneously with removing the inactive flag and therefore this will not
+				// initiate disconnection
+				ns.SetStateSub(node, nodestate.Flags{}, ppSetup.inactiveFlag, 0)
+			}
+		}
+		if oldState.Equals(ppSetup.inactiveFlag) && newState.Equals(ppSetup.inactiveFlag.Or(btSetup.priorityFlag)) {
+			ns.SetStateSub(node, ppSetup.inactiveFlag, nodestate.Flags{}, 0) // priority gained; remove timeout
+		}
+		if newState.Equals(ppSetup.activeFlag) {
+			// active with no priority; limit capacity to minCap
+			cap, _ := ns.GetField(node, ppSetup.capacityField).(uint64)
+			if cap > minCap {
+				cp.requestCapacity(node, minCap, 0, true)
+			}
+		}
+		if newState.Equals(nodestate.Flags{}) {
+			if c, ok := ns.GetField(node, clientField).(clientPeer); ok {
+				c.Disconnect()
+			}
+		}
+	})
+
+	ns.SubscribeField(btSetup.balanceField, func(node *enode.Node, state nodestate.Flags, oldValue, newValue interface{}) {
+		if newValue != nil {
+			ns.SetStateSub(node, ppSetup.inactiveFlag, nodestate.Flags{}, 0)
+			cp.lock.RLock()
```

Comment:
I think we should put the default price policies into the balance tracker. So that all the newly created balances will be assigned automatically, instead of doing here.

Reply:
done